torch/autograd/functional.py

```diff
@@ -0,0 +1,714 @@
+
+    Args:
+        func (function): a Python function that takes Tensor inputs and returns
+            a tuple of Tensor or a Tensor.
+        inputs (tuple of Tensor or Tensor): inputs to the function.
+        create_graph (bool, optional): If ``True``, the Jacobian will be computed in
+            a differentiable manner. Note that when strict if ``False``, the result can not
+            require gradients or be disconnected from the inputs.
+            Defaults to ``False``.
+        strict (bool, optional): If ``True``, an error will be raised when we detect we detect that some quantities
+            are independent of the element we want to differentiate it with. If ``False``, such quantity
+            will only be set to ``0`` which is the expected mathematical value.
+            Defaults to ``False``.
+
+    Returns:
+        Jacobian (Tensor or nested tuple of Tensor) if there are a single input
+            and output, this will be a single Tensor containing the Jacobian for the
+            linearized inputs and output. If one of the two is a tuple, then the Jacobian
+            will be a tuple of Tensors. If both of them are tuples, then the Jacobian will
+            be a tuple of tuple of Tensors where ``Jacobian[i][j]`` will contain the Jacobian
+            of the ``i``th output and ``j``th input and will have as size the sum of the size
+            of the corresponding output plus the size of the corresponding input.
+
+    Example::
+
+        >>> def exp_reducer(x):
+        ...   return x.exp().sum(dim=1)
+        >>> inputs = torch.rand(2, 2)
+        >>> jacobian(exp_reducer, inputs)
+        tensor([[[1.4917, 2.4352],
+                 [0.0000, 0.0000]],
+
+                [[0.0000, 0.0000],
+                 [2.4369, 2.3799]]])
+
+        >>> jacobian(exp_reducer, inputs, create_graph=True)
+        tensor([[[1.4917, 2.4352],
+                 [0.0000, 0.0000]],
+
+                [[0.0000, 0.0000],
+                 [2.4369, 2.3799]]], grad_fn=<ViewBackward>)
+
+        >>> def exp_adder(x, y):
+        ...   return 2 * x.exp() + 3 * y
+        >>> inputs = (torch.rand(2), torch.rand(2))
+        >>> jacobian(exp_adder, inputs)
+        (tensor([[2.8052, 0.0000],
+                [0.0000, 3.3963]]),
+         tensor([[3., 0.],
+                 [0., 3.]]))
+    """
+
+    tuple_inputs, inputs = _as_tuple(inputs, "inputs", "jacobian")
+    inputs = _grad_preprocess(inputs, create_graph=create_graph, need_graph=True)
+
+    outputs = func(*inputs)
+    tuple_outputs, outputs = _as_tuple(outputs, "outputs of the user-provided function", "jacobian")
+    _check_requires_grad(outputs, "outputs", strict=strict)
+
+    jacobian = tuple()
+    for i, out in enumerate(outputs):
+
+        jac_i = tuple([] for _ in range(len(inputs)))
+        for j in range(out.nelement()):
+            # Do this instead of the grad_output trick in `gradcheck` to avoid version counter issues
```

Comment:
What were the version counter issues?

Reply:
When we have `create_graph=True`, and `grad_output` would be saved when creating the graph in the backward, we cannot use a similar trick where we set a `1` at one coordinate one by one. Because the saved `grad_output` in the graph will then have an invalid version.